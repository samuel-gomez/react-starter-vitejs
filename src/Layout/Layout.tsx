import type { ComponentPropsWithoutRef, ReactNode } from 'react';
import withClassNameModifier, { type TwithClassNameModifier } from 'shared/hoc/WithClassNameModifier';
import Header from 'Layout/Header';
import Footer from 'Layout/Footer';
import TitleBar from 'Layout/TitleBar';
import Menu from 'Layout/Menu';
import A11yMenu from 'Layout/A11yMenu';

export type TLayout = {
  children?: ReactNode;
  className?: string;
  fullScreen?: boolean;
  disabled?: {
    header?: boolean;
    menu?: boolean;
    title?: boolean;
    footer?: boolean;
    a11yMenu?: boolean;
  };
  propsHeader?: ComponentPropsWithoutRef<typeof Header>;
  propsMenu?: ComponentPropsWithoutRef<typeof Menu>;
  propsTitle?: ComponentPropsWithoutRef<typeof TitleBar>;
  propsFooter?: ComponentPropsWithoutRef<typeof Footer>;
  propsA11yMenu?: ComponentPropsWithoutRef<typeof A11yMenu>;
  A11yMenuCmpt?: typeof A11yMenu;
  HeaderCmpt?: typeof Header;
  MenuCmpt?: typeof Menu;
  TitleBarCmpt?: typeof TitleBar;
  FooterCmpt?: typeof Footer;
} & TwithClassNameModifier;

export type TLayoutPage = TLayout & {
  titleBar?: string;
  title?: ReactNode;
};

const disabledDefault = {
  header: false,
  menu: false,
  title: false,
  footer: false,
  a11yMenu: false,
};

const Layout = withClassNameModifier(
  ({
    className,
    children,
    propsHeader,
    propsMenu,
    propsTitle,
    propsFooter,
    propsA11yMenu,
    fullScreen,
    disabled = disabledDefault,
    A11yMenuCmpt = A11yMenu,
    HeaderCmpt = Header,
    MenuCmpt = Menu,
    TitleBarCmpt = TitleBar,
    FooterCmpt = Footer,
  }: TLayout) => (
    <>
      {!disabled.a11yMenu && <A11yMenuCmpt {...propsA11yMenu} />}
      {!disabled.header && <HeaderCmpt {...propsHeader} fullScreen={fullScreen} />}
      {!disabled.menu && <MenuCmpt {...propsMenu} fullScreen={fullScreen} isVisible />}
      {!disabled.title && <TitleBarCmpt {...propsTitle} fullScreen={fullScreen} />}
      <main aria-label="Main Content" id="main-content" className={className}>
        {fullScreen ? children : <section className="container">{children}</section>}
      </main>
      {!disabled.footer && <FooterCmpt {...propsFooter} fullScreen={fullScreen} />}
    </>
  ),
  { defaultClassName: 'af-main' },
);

export default Layout;
