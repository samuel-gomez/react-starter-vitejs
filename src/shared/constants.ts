import packageJson from '../../package.json';

/** ***********************************************************************
 * FORM CONSTANTS
 ************************************************************************ */
export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_LENGTH_3 = 'Le champ doit faire au moins 3 caractères';
export const MSG_FORMAT = 'Format invalide';
export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

// patterns
export const patternEmail = /^[\w-+]+@([\w-+]+\.)+[\w-]{2,}$/;
export const patternPhoneNumber = /^[+0][\d ]*$/;
export const patternZipCode = /^\d{5}$/;
export const patternSiret = /^\d{14}$/;
export const patternSiren = /^\d{9}$/;
export const patternSiretSiren = /^\d{9,14}$/;

export const phoneNumberRule = {
  required: true,
  minLength: 4,
  maxLength: 20,
  pattern: patternPhoneNumber,
};

export const emailRule = {
  required: true,
  minLength: 6,
  maxLength: 60,
  pattern: patternEmail,
};

/** ***********************************************************************
 * API STATUS CONSTANTS
 ************************************************************************ */
export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

/** ***********************************************************************
 * API URL
 ************************************************************************ */
export const API_URL = {
  BASE: 'base',
  GITHUB: 'github',
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

/** ***********************************************************************
 * USER CONSTANTS
 ************************************************************************ */

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';
export const ALL = '';

export const PROFILS = [ADMIN, READER, WRITER, USER, ALL];

/** ***********************************************************************
 * COMMONS CONSTANTS
 ************************************************************************ */

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

/** ***********************************************************************
 * VERSION CONSTANT
 ************************************************************************ */

const VERSION_TOOLKIT = 'master';

/** ***********************************************************************
 * DEMO CONSTANTS
 ************************************************************************ */

export const STORYBOOK = 'https://axafrance.github.io/react-toolkit/latest/storybook/?path=/story/';
export const DESIGN_SYSTEM = 'https://axafrance.github.io/design-system/';
export const GITHUB = `https://github.com/AxaGuilDEv/react-toolkit/tree/v${packageJson.dependencies['@axa-fr/react-toolkit-all']}/packages/`;
export const GITHUB_API = `AxaGuilDEv/react-toolkit/${VERSION_TOOLKIT}/packages/`;
