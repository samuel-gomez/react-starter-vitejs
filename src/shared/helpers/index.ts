export { formatDate, isValidDate, setDate } from './formatDate';
export { default as isEmptyOrNull, isEmptyArray, isEmptyObject } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as emptyFunction } from './emptyFunction';
export { default as setRequiredFieldRule, setRequiredMessage } from './setRequiredFieldRule';
