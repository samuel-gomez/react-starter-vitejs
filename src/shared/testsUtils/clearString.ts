const cleanString = (str?: string) => str?.replace(/\s/g, '').replace(/\n/g, '').replace(/\t/g, '').trim();

export default cleanString;
