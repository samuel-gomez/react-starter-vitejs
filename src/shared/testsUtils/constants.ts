const MOCK_API_URL = { base: 'https://react-starter-api.vercel.app/api/', github: 'https://raw.githubusercontent.com/' };

export default MOCK_API_URL;

export const SCOPE_EDITOR = 'Edit props';
export const SCOPE_PREVIEW = 'af-accessibility';
export const SCOPE_CODE = 'af-accessibility-code';
