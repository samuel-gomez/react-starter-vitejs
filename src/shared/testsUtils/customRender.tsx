import { render, renderHook, type RenderOptions } from '@testing-library/react';
import EnvironmentProvider from 'App/EnvironmentProvider';
import FetchProvider from 'App/FetchProvider';
import NotificationProvider from 'App/NotificationProvider';
import QueryProvider from 'App/QueryProvider';
import UserProvider from 'App/UserProvider';
import { type ReactElement, type ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';
import MOCK_API_URL from './constants';

type TMockProvider = {
  [x: string]: Record<string, unknown | number | string> | string | boolean | null | number | Record<string, number | string>[];
};

const MockProviders =
  ({
    role = '',
    isEnabled = true,
    route = '/',
    oidcUser = { member_of: [`CN=${role}`], name: 'Samuel Gomez' },
    accessToken = 'accessToken',
    queryData,
    headers = {
      'Content-type': 'application/json; charset=UTF-8',
    },
  }: TMockProvider) =>
  ({ children }: { children: ReactNode }) => {
    const getAccessTokenFn = vi.fn().mockReturnValue(() => ({ accessToken }));
    const getUserInfosFn = vi.fn().mockReturnValueOnce(() => ({ oidcUser }));
    const useEnvFn = vi.fn().mockReturnValueOnce({
      envState: {
        environment: {
          apiUrl: MOCK_API_URL,
          fetchConfig: {
            headers,
          },
          oidc: {
            isEnabled,
          },
        },
      },
    });

    const queriesOptions = {
      retry: false,
      ...(queryData ? { queryFn: () => queryData } : {}),
    };

    return (
      <EnvironmentProvider useEnvFn={useEnvFn}>
        <UserProvider getUserInfosFn={getUserInfosFn}>
          <FetchProvider getAccessTokenFn={getAccessTokenFn}>
            <QueryProvider queriesOptions={queriesOptions}>
              <NotificationProvider>
                <MemoryRouter initialEntries={[`${route}`]}>{children}</MemoryRouter>
              </NotificationProvider>
            </QueryProvider>
          </FetchProvider>
        </UserProvider>
      </EnvironmentProvider>
    );
  };

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>, testMock: TMockProvider = { role: '' }) =>
  render(ui, { wrapper: MockProviders(testMock), ...options });

type TcustomRenderHook = (testMock?: TMockProvider) => typeof renderHook;

const customRenderHook: TcustomRenderHook = testMock => (hook, options) => renderHook(hook, { wrapper: MockProviders(testMock ?? {}), ...options });

// re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { axe } from 'jest-axe';

// override render method
export { customRenderHook, customRender as render };
