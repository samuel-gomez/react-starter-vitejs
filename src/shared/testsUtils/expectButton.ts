import { screen, within } from '@testing-library/dom';

type TexpectButton = {
  name: string;
  isQueryByRole?: boolean;
  beDisabled?: boolean;
  beInDoc?: boolean;
  parentLabel?: string;
};

const expectButton = ({ isQueryByRole = false, name, beDisabled = true, beInDoc = true, parentLabel = '' }: TexpectButton) => {
  const base = parentLabel ? within(screen.getByLabelText(parentLabel)) : screen;
  const button = isQueryByRole ? base.queryByRole('button', { name: RegExp(name) }) : base.getByRole('button', { name: RegExp(name) });
  if (beInDoc) {
    expect(button).toBeInTheDocument();
    if (beDisabled) {
      expect(button).toBeDisabled();
    } else {
      expect(button).not.toBeDisabled();
    }
  } else {
    expect(button).not.toBeInTheDocument();
  }
};

export default expectButton;
