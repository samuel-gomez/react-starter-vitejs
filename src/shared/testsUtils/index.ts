export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
