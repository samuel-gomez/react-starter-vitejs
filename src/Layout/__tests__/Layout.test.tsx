import { describe, it, expect } from 'vitest';
import { renderWithWrapperStaticRouter } from 'shared/testsUtils';
import Layout from '../Layout';

const defautlProps = {
  A11yMenuCmpt: () => <p>A11yMenuCmpt</p>,
  HeaderCmpt: () => <p>HeaderCmpt</p>,
  MenuCmpt: () => <p>MenuCmpt</p>,
  TitleBarCmpt: () => <p>TitleBarCmpt</p>,
  FooterCmpt: () => <p>FooterCmpt</p>,
};

describe('Layout', () => {
  it.each`
    fullScreen | disabled
    ${false}   | ${{ header: true }}
    ${false}   | ${{ title: true }}
    ${false}   | ${{ footer: true }}
    ${false}   | ${{ menu: true }}
    ${true}    | ${{ header: true, title: true, menu: true, footer: true }}
  `('Should render when disabled: $disabled', ({ disabled, fullScreen }) => {
    const { asFragment } = renderWithWrapperStaticRouter(<Layout disabled={disabled} fullScreen={fullScreen} {...defautlProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
