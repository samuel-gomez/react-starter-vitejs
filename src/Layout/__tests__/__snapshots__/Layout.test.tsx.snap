// Vitest Snapshot v1

exports[`Layout > Should render when disabled: { footer: true } 1`] = `
<DocumentFragment>
  <p>
    A11yMenuCmpt
  </p>
  <p>
    HeaderCmpt
  </p>
  <p>
    MenuCmpt
  </p>
  <p>
    TitleBarCmpt
  </p>
  <main
    class="af-main"
    id="main-content"
  >
    <section
      class="container"
    />
  </main>
</DocumentFragment>
`;

exports[`Layout > Should render when disabled: { header: true } 1`] = `
<DocumentFragment>
  <p>
    A11yMenuCmpt
  </p>
  <p>
    MenuCmpt
  </p>
  <p>
    TitleBarCmpt
  </p>
  <main
    class="af-main"
    id="main-content"
  >
    <section
      class="container"
    />
  </main>
  <p>
    FooterCmpt
  </p>
</DocumentFragment>
`;

exports[`Layout > Should render when disabled: { header: true, title: true, â€¦(2) } 1`] = `
<DocumentFragment>
  <p>
    A11yMenuCmpt
  </p>
  <main
    class="af-main"
    id="main-content"
  />
</DocumentFragment>
`;

exports[`Layout > Should render when disabled: { menu: true } 1`] = `
<DocumentFragment>
  <p>
    A11yMenuCmpt
  </p>
  <p>
    HeaderCmpt
  </p>
  <p>
    TitleBarCmpt
  </p>
  <main
    class="af-main"
    id="main-content"
  >
    <section
      class="container"
    />
  </main>
  <p>
    FooterCmpt
  </p>
</DocumentFragment>
`;

exports[`Layout > Should render when disabled: { title: true } 1`] = `
<DocumentFragment>
  <p>
    A11yMenuCmpt
  </p>
  <p>
    HeaderCmpt
  </p>
  <p>
    MenuCmpt
  </p>
  <main
    class="af-main"
    id="main-content"
  >
    <section
      class="container"
    />
  </main>
  <p>
    FooterCmpt
  </p>
</DocumentFragment>
`;
